sort the edits in descending order

ananthakumaran opened the pull request:
fixes #288

> Applying the edit instructions in reverse to file will result in correctly reformatted text.

One of the comments in TypeScript repo says the edits should be applied
in reverse order. But, I am not sure all the plugins follows this? Doing
a manual sort does seem to fix the tslint issue. VS code applies the fixes
correctly, but I can't able to find the logic. Atom[1] also manually
sorts the edits.

1: https://github.com/TypeStrong/atom-typescript/blob/master/lib/main/pluginManager.ts#L349-L352

ananthakumaran commented:
@jupl could you try this branch?

jupl commented:
I can still replicate the issue unfortunately. :(

jupl commented:
https://github.com/jupl/tide-thingy

I can fix the `var` keyword with `tide-fix`, but attempting to fix curly spacing with `s` or `d` gets wacky. It works as expected in VSCode.

ananthakumaran commented:
Thanks for the example, it was helpful in debugging. I have added two more fixes
1) fixes for same file was sent as two different code-edits, I have to do a group by filename to fix this
2) we apply format-region on the region where the new code-edits are applied. This was a hack to prevent code-edit messing up the format, but in your case, tide-format was adding the spaces back. I have removed the tide-format.

I am kind of not sure about merging these changes. AFAIK, these things are not thoroughly documented anywhere, so I am kind of not sure what is the correct fix and who is at fault here.

ananthakumaran commented:
```typescript
    /** Apply each textChange in order, updating future changes to account for the text offset of previous changes. */
    function forEachTextChange(changes: ReadonlyArray<ts.TextChange>, cb: (change: ts.TextChange) => void): void {
        // Copy this so we don't ruin someone else's copy
        changes = JSON.parse(JSON.stringify(changes));
        for (let i = 0; i < changes.length; i++) {
            const change = changes[i];
            cb(change);
            const changeDelta = change.newText.length - change.span.length;
            for (let j = i + 1; j < changes.length; j++) {
                if (changes[j].span.start >= change.span.start) {
                    changes[j].span.start += changeDelta;
                }
            }
        }
    }
```

from this logic it looks like applying the changes in reverse order should work, as there won't be any more span after the current one.

ananthakumaran commented:
@jupl could you test it once more
